File: packages/gbnf-experiment/data/20260909T180307Z_280f3890/ported_implementation/tests/validation/validate-grammar.test.ts
// Ported from the generated Python suite at tests/python.
import { describe, expect, it } from 'vitest';

import { GBNF, GrammarParseError } from '../../src/index.js';

describe('validate grammar', () => {
  const validGrammars: [grammar: string][] = [
    ['root ::= "foo"'],
    ['root ::= "foo" | "bar"'],
    ['root ::= ("foo" | "bar")'],
    ['root ::= ("foo" | "bar")?'],
    ['root ::= ("foo" | "bar")*'],
    ['root ::= ("foo" | "bar")+'],
    ['root ::= [a-z]'],
    ['root ::= [a-zA-Z]'],
    ['root ::= [a-zA-Z0-9]'],
    ['root ::= [a-zA-Z0-9]*'],
    ['root ::= [a-zA-Z0-9]?'],
    ['root ::= [a-z]+'],
    ['root ::= [a-zA-Z0-9]+'],
    ['root ::= ([a-zA-Z0-9])*'],
    ['root ::= ([a-zA-Z0-9])?'],
    ['root ::= ([a-zA-Z0-9])+'],
    ['\n  root ::= foo\n  foo ::= "foo"'],
    ['\n  root ::= foo\n  foo ::= "foo" | "bar" | ([a-z])?\n  '],
  ];

  it.each(validGrammars)('parses a grammar (case %#)', grammar => {
    expect(() => GBNF(grammar)).not.toThrow();
  });

  const invalidGrammars: [grammar: string, errorPos: number, errorReason: string][] = [
    ['', 0, 'No rules were found'],
    ['root = "foo"', 5, 'Expecting ::= at 5'],
    ['root ::= foo\nfoo := "foo"\n  ', 17, 'Expecting ::= at 17'],
    ['root ::= foo', 9, 'Undefined rule identifier "foo"'],
    ['root ::= foo\nbar ::= "bar"\n  ', 9, 'Undefined rule identifier "foo"'],
    ['root ::= foo\nfoo ::= baz\nbar ::= "bar"\n  ', 21, 'Undefined rule identifier "baz"'],
    ['root ::= foo ::= bar', 13, 'Expecting newline or end at 13'],
    ['root ::= ([a-z]\nfoo ::= "foo"\n  ', 20, 'Expecting \')\' at 20'],
  ];

  it.each(invalidGrammars)(
    'reports an error for an invalid grammar (case %#)',
    (grammar, errorPos, errorReason) => {
      const expected = new GrammarParseError(grammar, errorPos, errorReason);
      expect(() => GBNF(grammar)).toThrow(GrammarParseError);
      try {
        GBNF(grammar);
        expect.unreachable('expected GBNF to throw');
      } catch (err) {
        expect((err as GrammarParseError).message).toBe(expected.message);
      }
    },
  );
});

File: proposed-projects/round-trip-experiment/reverse/20260909T125752Z_09b2c811/20260910T072624Z_d6bd66eb/ported_implementation/tests/validation/validate_grammar_test.py
import pytest
from gbnf import GBNF, GrammarParseError


def describe_validate_grammar():
    @pytest.mark.parametrize(
        ("grammar"),
        [
            '''root ::= "foo"''',
            '''root ::= "foo" | "bar"''',
            """root ::= ("foo" | "bar")""",
            """root ::= ("foo" | "bar")?""",
            """root ::= ("foo" | "bar")*""",
            """root ::= ("foo" | "bar")+""",
            """root ::= [a-z]""",
            """root ::= [a-zA-Z]""",
            """root ::= [a-zA-Z0-9]""",
            """root ::= [a-zA-Z0-9]*""",
            """root ::= [a-zA-Z0-9]?""",
            """root ::= [a-z]+""",
            """root ::= [a-zA-Z0-9]+""",
            """root ::= ([a-zA-Z0-9])*""",
            """root ::= ([a-zA-Z0-9])?""",
            """root ::= ([a-zA-Z0-9])+""",
            '''
  root ::= foo
  foo ::= "foo"''',
            """
  root ::= foo
  foo ::= "foo" | "bar" | ([a-z])?
  """,
        ],
    )
    def test_it_parses_a_grammar(grammar):
        GBNF(grammar)

    @pytest.mark.parametrize(
        ("grammar", "error_pos", "error_reason"),
        [
            ("", 0, "No rules were found"),
            ('root = "foo"', 5, "Expecting ::= at 5"),
            (
                """root ::= foo
foo := "foo"
  """,
                17,
                "Expecting ::= at 17",
            ),
            ("root ::= foo", 9, 'Undefined rule identifier "foo"'),
            (
                """root ::= foo
bar ::= "bar"
  """,
                9,
                'Undefined rule identifier "foo"',
            ),
            (
                """root ::= foo
foo ::= baz
bar ::= "bar"
  """,
                21,
                'Undefined rule identifier "baz"',
            ),
            ("root ::= foo ::= bar", 13, "Expecting newline or end at 13"),
            (
                """root ::= ([a-z]
foo ::= "foo"
  """,
                20,
                "Expecting ')' at 20",
            ),
        ],
    )
    def test_it_reports_an_error_for_an_invalid_grammar(
        grammar, error_pos, error_reason
    ):
        expected = GrammarParseError(grammar, error_pos, error_reason)
        with pytest.raises(GrammarParseError) as e:
            graph = GBNF(grammar)
        e = e.value
        assert e == expected
